### DISCUSSION TITLE
Initial setup questions and feedback

### DISCUSSION THREAD

**Discussion by netmindz** [netmindz/balboa_GL_ML_spa_control#24] [Q&A]
Please use this discussion for any questions about what control units are supported, the current feature set, the hardware needed, Home Assistant integration etc

https://github.com/netmindz/balboa_GL_ML_spa_control/wiki/Install-Guide

**Comment by caldog20** [comment]
Question on the wiring setup.

I am using an ESP32. The RS485 TTL converter I purchased claims to support both 5v and 3.3v logic levels at the same time. 

I did go ahead and buy bi-directional 3.3v/5v logic lever shifters. Are you using a logic level shifter for the rs485 serial connections and for Pin5 or stepping down Pin5 another way?

I have a buck converter that produces 5V I can use to power the esp32 through the USB port, with power incoming from the tub.

Also, from the install guide, I don't believe we do anything with pins 2 or 4 correct?

Thanks,

**Reply by netmindz** [reply]
At first I powered the RS485 adapter as 3.3v and used a level shifter for pin 5 (panel select). However, I think the level shifter was introducing a delay, as I had much better success when I risked applying 5v to the ESP32 and for both me and @tmjo we found we could not send any data to the RS845 unless it was powered by 5v.

While the output of the esp32 is only 3.3v it appears to be 5v tolerant to some extent (I have killed a few during this project). I'm more of a software developer than electronics engineer, so might be better way to get lag-free 3.3v using voltage divider resistor setup or similar, but personally at the moment I'm just banging 5v from pin 5 into the esp32 without issue

**Comment by caldog20** [comment]
One other issue I am wondering If I will have is that my original ML700 panel is toast. Therefore, I won't be able to sniff a UUID of the panel if that is needed. I imagine if we can calculate the right CRC value based on the UUID it won't matter? Not sure if the board actually verifies the UUID of the panel and compares.

**Reply by netmindz** [reply]
I'm pleased to report that @tmjo has been able to drive his control unit using the uid value from my top panel, so no need to change  to your own. So long as we have FB commands with valid CRC we are all good, by removing his uid values and putting all the sixes as placeholder has just confused things a bit really

**Comment by StumpyRob** [comment]
Hi Guys, 

All the kit has arrived so I'm ready to have a play this weekend! First off I want to thank you guys for al your work. I will try to spread the word as much as possible about this. - I'm surprised it hasn't got more exposure! 

I have a few questions: 
1. I have used Platformio to upload the code to my ESP32 (ESP32-WROOM-32D) I have a successful upload for the wemos_d1_mini32 but a failure for remote_wemos_d1_mini32 and d1_mini_lite. I'm presuming that's normal behaviour? - the code has been designed for multiple boards and the correct one has been uploaded for my ESP32? 
(First time I have used Platformio and I'm quite impressed by it!)
<a href="https://ibb.co/CMhKqcv"><img src="https://i.ibb.co/pRJjg7Z/Screenshot-2023-06-20-224630.png" alt="Screenshot-2023-06-20-224630" border="0"></a>

2. In the installation instructions, after upload of code it states _"Then go to that address in the status page"_ - I can see the IP that the ESP has received form DHCP, I can ping it and I can telnet to it. Does the ESP host its own webpage? Am I expected to see something if I browse to it on port 80/81?

3. Finally, I can see in HA that the ESP32 is trying to connect to the MQTT mosquito broker. - awesome! However, I'm getting the error: _"New connection from (IP ADDRESS)  on port 1883. Error: received null username or password for unpw check"_ 
Is there a way to add credentials to the ESP32 for the broker? - I don't think mosquito allows anon. credentials. If not, can you advise what setup you are using? 
<a href="https://ibb.co/M1NDD9Y"><img src="https://i.ibb.co/09JjjQg/Screenshot-2023-06-20-225329.png" alt="Screenshot-2023-06-20-225329" border="0"></a><br />

I did also had a question about pin5 not being directly connected also however, @caldog20 has already asked about that. 👍 

I would love help contribute to this. Once I get it going I would be happy to write a low level guide for users like myself!

**Reply by netmindz** [reply]
Thanks for the feedback

1) yeah you only need to run the "upload and monitor" for a single target

2) yeah there is a status web page on the normal port 80

3) you need to add a definite for BROKER_USERNAME, and BROKER_PASSWORD after the BROKER_ADDR if you are using auth

**Reply by StumpyRob** [reply]
Thanks for the reply! Success! - well kinda. 

I cant for the life of me get the HTTP status page to work HOWEVER: it's in Home assistant now and I can see the sensors and controls in MQTT! 👍 

I currently have the ESP32 sat on my workbench, nothing is soldered to the board yet so obviously, I'm not getting any feedback from the tub. - Next job is to solder everything up and get it plugged into the GL2000 I will let you know how I get on.

Just a thought: I doubt it very much but do I need the board to be plugged into the tub to get a response via HTTP? I'm guessing not, I presume I would still get the web page but obviously with no tub data.

**Reply by netmindz** [reply]
I have the webserver disabled in my work in progress branches but should be working on the main branch

**Reply by netmindz** [reply]
Webserver and websocket server both enabled again now

**Reply by StumpyRob** [reply]
Youre correct, that was the issue. All working now! 
<a href="https://imgbb.com/"><img src="https://i.ibb.co/NmrrdsL/Screenshot-2023-06-24-162009.png" alt="Screenshot-2023-06-24-162009" border="0"></a>

Ok , All I am waiting for is my Molex connector to turn up from the USA. No Idea why the 300mm versions are nowhere to be found in the UK. Once it turns up I will let you know how I get on.

Thanks so much for your help! 👍

**Reply by netmindz** [reply]
How long is the lead time? I might have a spare

**Reply by StumpyRob** [reply]
> How long is the lead time? I might have a spare

Thank you for the offer. I should be ok though. I have paid for UPS express so I'm hoping it will be with me by the end of the week. 🙂

**Reply by StumpyRob** [reply]
Hi Will, 

Ok the cable arrived! I did some soldering last night - finished at 10PM and obviously couldn't not plug it in and test it even if it meant working in low light. 
It all lights up and there was no burning silicone smell or smoke so that's a good start so far!
<a href="https://ibb.co/b1jp5fh"><img src="https://i.ibb.co/MnyqkXT/Microsoft-Teams-image-5.jpg" alt="Microsoft-Teams-image-5" border="0" /></a>

Unfortunately, I couldn't get any output from the tub.
It's likely myself misinterpreting the instructions. Here is how I have wired up the ESP32: 
<a href="https://ibb.co/rvs0n5Z"><img src="https://i.ibb.co/yVSPLg8/Screenshot-2023-06-29-155210.png" alt="Screenshot-2023-06-29-155210" border="0" /></a>

Can you advise if this is correct?

I was looking at the code and I can see that RX PIN 19 and TX PIN 23 are defined as well as PIN 22:

```
#ifdef ESP32
#define tub Serial2
#define RX_PIN 19
#define TX_PIN 23
#define RTS_PIN 22 // RS485 direction control, RequestToSend TX or RX, required for MAX485 board.
```

Looking at it now I am presuming that I should be wiring to PIN 23 and PIN 19 instead of "RX" and "TX" on the board. I just want to confirm before I go and resolder it :)

**Reply by StumpyRob** [reply]
Ok quick update, I have resoldered to pin 19 and 23. I have plugged in and I have the same result. I can get to the webpage but no info from the tub. That said, when I telnet onto the ESP I am getting an output: 

`=~=~=~=~=~=~=~=~=~=~=~= PuTTY log 2023.06.29 16:26:58 =~=~=~=~=~=~=~=~=~=~=~=
Hello 0
fa14333230430000610000008000101b5900000000007e
ae0d000000000000000000000000002d
ae0d000000000000000000000000002d
fa14333230430000610000008000101b5900000000007e
fa14333230430000610000008000101b5900000000007e
ae0d010000000000000000000000005a
ae0d010000000000000000000000005a
fa14333230430000610000008000101b5900000000007e
fa14333230430000610000008000101b5900000000007e
ae0d02000000000000000000000000c3
ae0d02000000000000000000000000c3
fa14333230430000610000008000101b5900000000007e
fa14333230430000610000008000101b5900000000007e
ae0d03000000000000000000000000b4
ae0d03000000000000000000000000b4
fa14333230430000610000008000101b5900000000007e
fa14333230430000610000008000101b5900000000007e
ae0d000000000000000000000000002d
ae0d000000000000000000000000002d
fa14333230430000610000008000101b5900000000007e
fa14333230430000610000008000101b5900000000007e
ae0d010000000000000000000000005a
ae0d010000000000000000000000005a
`

If you could confirm my wiring that would be great. - I can start looking elsewhere for the issue if I know I haven't messed up the soldering. :)

**Reply by netmindz** [reply]
Yeah, you were right to swap to the new pins not use RX and TX

The data you can see on the telnet output is valid Balboa GL data.

Are you using the main branch? Is your tub set for C or F for it's temperature?

**Reply by netmindz** [reply]
If your can also send the output from the serial monitor too

**Reply by netmindz** [reply]
I've just pushed the newer code that supports read-write control of the tub to `main` btw

**Reply by netmindz** [reply]
Can you add a new diagram with the connections to the right pins please @StumpyRob and I will add to the docs

**Reply by StumpyRob** [reply]
Awesome! Thanks for the reply.

I will get a diagram over to you later on tonight. 

The tub is in centigrade I will have a play probably tomorrow with the new code and let you know how I get on. I will also take a look at the serial output too. 🙂

**Reply by StumpyRob** [reply]
Here is the edited version. Glad I could contribute in some way - even if it is just a crappy diagram! 🙂
<a href="https://ibb.co/2hfP1Nw"><img src="https://i.ibb.co/Gpbk4TX/Screenshot-2023-06-29-220058.png" alt="Screenshot-2023-06-29-220058" border="0" /></a>

**Reply by netmindz** [reply]
Every bit helps. Guide updated

**Comment by caldog20** [comment]
Just a quick update. My supplies came in, and I was able to do some testing. 

The transformer on my board fried so I had to order a replacement to fix it. But until that comes in, I was able to find another 120V  to 12V AC transformer to connect up to the board to power the low voltage components just for testing. I was able to communicate with the board from my esp32 and also send commands. I was able to toggle the light and jets. I was also able to get the temperature. 

As soon as my new transformer comes in, I will have the board back in the tub so I can see what it shows while actually running the tub.

**Reply by netmindz** [reply]
Great to hear your feedback. What's the config of your tub?

Dual jets? Single speed or dual speed pumps?

I'm guessing if you are using 120V then you might also normally be running your tub with the temp shown in F rather than Centigrade.  Are you just seeing the raw temp come through so Home Assistant says C but is actually F?

**Reply by caldog20** [reply]
My tub is 240V, but it doesn't use a neutral. Just 2 hot wires and a ground. It has an on-board 120V transformer for the lower voltage components. 

Yeah I am in the US so its set to F. I am not using home assistant yet, I was just decoding the values from the messages I was seeing and it uses 0x46 instead of 0x43 after the temp values. 

I have 2 dual speed pumps, no circulation pump.

**Reply by netmindz** [reply]
You will need to swap to 46 then on line 525 ish where there is `result.substring(10, 12) == "43"` and ` targetTemp.setUnitOfMeasurement("°F");`

**Reply by netmindz** [reply]
and uncomment the PUMP1_DUAL_SPEED and PUMP2_DUAL_SPEED

Will probably need some other tweaks to detect when pump1 is in low due to cleaning

**Reply by netmindz** [reply]
After the temp tweak, were you seeing the right status information shown?

**Reply by caldog20** [reply]
I finally got my new transformer and got it soldered onto the board. Going to test it today and then install board back in tub for a real test. I will report back this evening hopefully if everything goes well.

**Reply by caldog20** [reply]
I connected up the ESP32 and have some issues with the communication. I can sometimes get some proper messages but it cannot send commands. However, if I hook up my ATMega2560 which uses 5v logic it can send commands properly. 

I am not sure if I need to put a logic level shifter between the ESP32 RX/TX and the RS485 board, or if my ESP32 is just flakey. Will do some more testing today.

**Reply by netmindz** [reply]
The VCC supply to the RS485 adapter must be 5v. I've had issues with delays when using level shifters, so I'm using without

**Reply by caldog20** [reply]
I think it may be the way I am receiving the messages. I am not really ever matching the logic to handle the message or send a command. It seems like I am getting more data at once that expected. For instance, I never match this line: `if(msgLength == 0 && result.length() == 2) `

But I am receiving the full messages

**Reply by netmindz** [reply]
Which branch are you using? Do you have pin5 connected?

**Reply by caldog20** [reply]
send-command-v2. Yes I am receiving messages from the tub, FA and AE messages. 

If I put a print statement here under this loop 
```
      for (int i = 0; i < len; i++) {
        if (buf[i] < 0x10) {
          result += '0';
        }
        result += String(buf[i], HEX);
      }
```
      
I get output like this

```
ae
ae0d
ae0d01
ae0d0100
ae0d010000
ae0d01000000
ae0d0100000000
ae0d010000000000
ae0d01000000000000
ae0d0100000000000000
ae0d010000000000000000
ae0d01000000000000000000
ae0d0100000000000000000000
ae0d010000000000000000000000
ae0d01000000000000000000000000
ae0d010000000000000000000000005a
fa
fa14
fa1420
fa142039
fa14203933
fa1420393346
fa142039334601
fa14203933460110
fa1420393346011029
fa142039334601102900
fa14203933460110290004
fa1420393346011029000400
fa142039334601102900040080
fa14203933460110290004008002
fa14203933460110290004008002ff
fa14203933460110290004008002ffff
fa14203933460110290004008002ffff5d
fa14203933460110290004008002ffff5d00
fa14203933460110290004008002ffff5d0000
fa14203933460110290004008002ffff5d000000
fa14203933460110290004008002ffff5d00000000
fa14203933460110290004008002ffff5d0000000000
fa14203933460110290004008002ffff5d0000000000b8
```

Which to me it is working as intended and should be matching the logic you wrote and parsing the message properly. I will keep troubleshooting.

**Reply by caldog20** [reply]
One thing though, I don't have any topside panels connected at all. Could this be a cause since there is not another panel on the bus to send replies for the board to process?

**Reply by caldog20** [reply]
I think my ESP is just bad. Going to order a few more and retry. I can use most commands manually though my Mega2560 and the logic for reading messages from the send-command-v2 branch works fine on this board, telling me my soldering and wiring is correct.

**Reply by netmindz** [reply]
Might be worth doing a quick zoom call so I can see exactly what results you are seeing

**Reply by caldog20** [reply]
What timezone are you in? We could do that. To be honest this board is like 4 years old and I probably damaged it somewhere along the way. I also have a few ESP8266 laying around I could try with, but their wifi sucks. 

I am looking around now for a new ESP32 board to buy that will be a good fit for this. My current board is ESP32 Devkit v1 ESP32-WROOM-32, which is nice because it can run at 240Mhz. But it's microUSB connector is almost shot. 

It would be really nice to get a USB-C board.

**Reply by netmindz** [reply]
I'm in the UK

**Reply by caldog20** [reply]
I figured it out. The ESP32 I was using is bad. I borrowed one until my new ones arrive. It's working fine now.

**Reply by netmindz** [reply]
tmjo also has dual-speed, but looks like my code isn't working quite right for that yet, I'm assuming it's like my previous tub where you press the jets button to cycle between off - low - high and therefore it's just sending the same command.

I look forward to a PR from you for any C vs F tweaks :)

**Reply by caldog20** [reply]
Sure, I can PR that for displaying proper temp based on the C/F byte value. 

Yes, for dual speed sending the same command cycles between Off/Low/High. I am currently using it and it works for both pumps. Glad I have everything hooked up now so I can actually start contributing!

**Reply by netmindz** [reply]
Possibly just use something based on this  https://github.com/netmindz/balboa_GL_ML_spa_control/pull/10 ?

Just to be clear - you have dual speed pumps and control from HA is working for both and at the correct speed?

**Reply by caldog20** [reply]
I will take a look! I want to go out though and switch my dipswitch to C temp and make sure the byte value being read in #10 is still F on my system, just incase there is a difference. I haven't done much with HA but I was thinking that if we check the byte for C/F in a status message we could then dynamically update `targetTemp.setUnitOfMeasurement("°F")` and `temp.setUnitOfMeasurement("°F")` so users wouldn't have to make changes if their tub is using either, if that's okay with everyone. But if we can't dynamically update the unit of measurement we could just have 2 temp values for current/target similar to #10.

Yeah, both of my pumps are dual speed. The speed setting for both pumps is currently working from HA for the most part. However, I think we may need to adjust it a little. For instance, if you are in High speed, and try to go to medium, _sometimes_ it turns it off. This is similar to the bug with the Mode setting. I went over the formula for the mode setting and it seems we are sending the right number of button presses, but I think it's possibly sending them too fast or too slow to be picked up. I planned on doing more testing on that and try to adjust. That would fix both the mode selection and make the pump speed selection more reliable.

**Reply by netmindz** [reply]
Single entity with the correct unit of measurement is fine. The original pr just added as second so we could compare. As it's a while back, I can't remember if I spotted we were deliberately rounding to 0.5 degree. Would be nice to see the most accurate value in HA

**Reply by tmjo** [reply]
I ran a pump test for my setup. Going to medium speed seems fine, but for me off and high are not working properly. When selecting high, it goes to high for a very brief second, but then stops. Got a serial log and there are two "yays" in there, even though code sent only once. I believe it could be that the HA state didn't get time to update, saw some comment in HA ARdunio that this could happen, but need to study it better.

```
---- Opened the serial port COM7 ----
onPumpSwitchStateChanged Pump1 1
Sending fb060343060006f9a2 - 1 times
message sent : 40
message sent : 40
YAY: command response : 40
onPumpSwitchStateChanged Pump1 2
Sending fb060343060006f9a2 - 1 times
message sent : 40
message sent : 40
message sent : 40
message sent : 40
message sent : 40
YAY: command response : 40
onPumpSwitchStateChanged Pump1 1
Sending fb060343060006f9a2 - 1 times
message sent : 40
message sent : 40
YAY: command response : 40
YAY: command response : 40
onPumpSwitchStateChanged Pump1 0
Sending fb060343060006f9a2 - 2 times
message sent : 40
message sent : 40
message sent : 40
message sent : 40
message sent : 40
YAY: command response : 40
message sent : 40
message sent : 40
message sent : 40
YAY: command response : 40
onPumpSwitchStateChanged Pump1 0
Sending fb060343060006f9a2 - 2 times
message sent : 40
message sent : 40
message sent : 40
message sent : 40
message sent : 40
YAY: command response : 40
message sent : 40
message sent : 40
message sent : 40
message sent : 40
YAY: command response : 40
onPumpSwitchStateChanged Pump1 2
Sending fb060343060006f9a2 - 1 times
message sent : 40
message sent : 40
message sent : 40
message sent : 40
message sent : 40
YAY: command response : 40

```

**Reply by netmindz** [reply]
Yeah, the double yay looks wrong. If I'm removing more than just the command that actually triggered the change, then we will miss a button press, but if there was only one in the buffer, then how can I remove 2? Something isn't right here.

The resending of commands until we see a change is a bit of a hack, I wonder if @caldog20 might be able to help remove that hack 

See https://github.com/netmindz/balboa_GL_ML_spa_control/issues/32

**Reply by caldog20** [reply]
I will take a look and try to submit a PR this weekend. I have to work late today but should have time tomorrow to do some testing on it.

**Reply by caldog20** [reply]
Sorry had family stuff going on this weekend. But good news, I got my replacement ESP32's in tonight so I can give back the one I borrowed. I will hook that up tomorrow and see if I can't get some things worked out with the command sending. Sorry for the delay.

**Reply by netmindz** [reply]
No worries fella

**Reply by netmindz** [reply]
Might be as simple as increasing the delay time till a single send of the command is reliable , but probably better to investigate the hardware flow control option https://github.com/netmindz/balboa_GL_ML_spa_control/issues/32

**Reply by tmjo** [reply]
CTS/RTS would be best, but could it be a solution to run interrupts when detecting PIN_5 going low to improve the timing?

**Reply by netmindz** [reply]
Given CTS/RTS is specifically listed as being used for half-duplex stuff like RS485, it is definitely an idea worth exploring. I've not really found much info on it as hardware flow control is such an old thing, so I can't confirm what we see on the scope. What surprises me is how it seems to toggle back and forth and a fixed clock rather than relating to if the controller is actually sending any data or not.

I did experiment a bit with the use of interrupts before I found that the level-shifter was adding a delay. If I remember correctly , then I could use that to trigger the read of the FA message, but if I tried to write then it broke stuff as you can't/shouldn't write serial data during the interrupt

**Reply by caldog20** [reply]
Are you thinking that a couple of the pins on the board may be used for flow control?

**Reply by caldog20** [reply]
Been doing some testing today. I have both the pump select and mode select switches working. I think we were just sending the commands too fast. For instance, when changing the mode select or the pump select, the first command is processed, but the rest are dropped. Typically, I think that the panel is sending empty replies between the button presses. If you think in terms of a human using the panel, there is a delay between pressing the buttons even if it's small.

Also, I have removed resending of commands and instantly dequeue them after they are sent. I haven't had any issues with needing to resend a command before getting a response, and none of the basic commands like up/down/light/mode button are dropped, so I think we are sending the command at the proper time. The delay in microseconds could probably stay incase others need to tune for different boards running at different frequencies, but I don't think the issue here is with that piece. 

For now, I have placed some empty commands between the actual commands when doing mode/pump selection changes. This seems to provide enough time for the board to process the commands between entering menu's etc. I am going to keep testing and see if there is a more elegant way to solve this. 

One thing I was thinking, we could define a command struct, similar to this:

```
struct Command {
  String command;
  int emptyCount; // basically a delay count
}
```
and use this in the HA callback functions, this way it's more clean, but can queue up empty commands more easily. Otherwise going off of a simple delay number might be difficult because of the place we are calling `sendCommand()` and the conditions around when it's triggered.

Anyways, I will keep trying to tune it. Let me know what you think.

**Reply by caldog20** [reply]
Also from my research, these boards don't use hardware flow control. Funnily enough, the board and top panels use the good old max485 IC for rs485 tranceivers. The flow control is usually done in software, or by some other means. This makes sense when you think about the pin5 for each panel. This is basically a flow control mechanism in itself, to allow the panel to know when to read or reply. Otherwise it will use the messages to determine flow control (when to send). I could be wrong but this is what I am leaning towards anyways.

**Reply by netmindz** [reply]
Interesting that you are getting better success rate than me. I see a few sends before my light will toggle. I have to admit though that my current physical setup is less than ideal as it's gone through a few iterations so I have wires stuffed into female headers where previous rs485 adapter boards were plugged in etc. It could do with a rebuild.

I could well be entirely wrong about pin 2 and 4. They must be for something, maybe I'm just desperate to know what 😅

Pin5 doesn't tell you when it's ok to transmit, but yes the protocol allow for a time period after the FA is sent by the controller for the top panel to send an FB so a delay is ok, but as an esp32 is more complicated than what they are using, we have various async stuff that the original boards don't which might make timing a little harder for us.

With regards to the command strut, this would only be needed if we need a different number of idle commands for different actions. 

Would be really interesting if someone could open up one of the panels and trace on the pcb where if anywhere pin 2 and pin 4 go

**Comment by caldog20** [comment]
I will take a look! I want to go out though and switch my dipswitch to C temp and make sure the byte value being read in #10 is still F on my system, just incase there is a difference. I haven't done much with HA but I was thinking that if we check the byte for C/F in a status message we could then dynamically update `targetTemp.setUnitOfMeasurement("°F")` and `temp.setUnitOfMeasurement("°F")` so users wouldn't have to make changes if their tub is using either, if that's okay with everyone. But if we can't dynamically update the unit of measurement we could just have 2 temp values for current/target similar to #10.

Yeah, both of my pumps are dual speed. The speed setting for both pumps is currently working from HA for the most part. However, I think we may need to adjust it a little. For instance, if you are in High speed, and try to go to medium, _sometimes_ it turns it off. This is similar to the bug with the Mode setting. I went over the formula for the mode setting and it seems we are sending the right number of button presses, but I think it's possibly sending them too fast or too slow to be picked up. I planned on doing more testing on that and try to adjust. That would fix both the mode selection and make the pump speed selection more reliable.

**Reply by netmindz** [reply]
My earlier code to try changing between standard and eco added space between the commands by sending idle commands. Might be worth a try adding an idle after each jet button press

**Reply by tmjo** [reply]
Yes, that is my thinking too. We see the original panels seems to send "empty" commands basically all the time. Perhaps there is some kind of reason for that. I was supposed to test during the weekend, but got caught up in other stuff that I had to take care of, so better luck this week perhaps.

**Reply by netmindz** [reply]
Yeah, I'm unsure what the point of that is, given the pin5 panel select time is fixes, so it's not like you are preventing blocking waiting for timeout for response from a panel where none might even be connected

**Comment by tmjo** [comment]
I'm seeing similar results as @caldog20, i.e. removing the resending of commands and adding some "padding" with the emty commands (typically 4) to have my commands act the way they should. Then pump, light and temp control is ok. I got some issues with the mode change (which used to work), but probably just some minor stuff I didn't see.

I will test with sending single commands on each PIN5 low. Depending on the speed of our ESPs it could be we're sending several commands on every PIN5 low, and perhaps that is not the intention? Perhaps it should be a single command.

**Reply by caldog20** [reply]
I can push what I have up to my fork if you want to take a look at see what is working for me. Mode/pumps/etc are currently working properly, but that may not be true for others. But it could at least be a reference. Mind the pin changes/etc since I use a different board. The sendCommand(), delayTime, and the callbacks for the HA button/switch is mainly what has changes. Let me know if any of it helps.

main branch here: https://github.com/caldog20/balboa_GL_ML_spa_control/tree/main

**Reply by caldog20** [reply]
Also there was a conflict I kept seeing between the Mode button and the Mode Select. It seemed there was a possible ID conflict and when you would toggle the mode select, it would send one extra command as if a button press was done in HA. Once I changed the ID of the Mode Select switch in the code, the extra un-wanted Mode command stopped sending. This was causing bugs with the mode select. So it is named differently in my fork.

**Reply by caldog20** [reply]
For instance, if changing the mode from Standard to Eco, it should have send these commands:

`Mode` -> `Down` -> `Mode`

but it would send `Mode` ->  `Mode` -> `Down` -> `Mode`  which would end up cancelling out the selection.

**Reply by caldog20** [reply]
Technically this shouldn't have happened since these are different HA entities, I know, but I tested it multiple times and the behavior was consistent.

**Reply by tmjo** [reply]
Thanks, that's sounds a lot like what I saw too. I'll verify.

**Reply by netmindz** [reply]
Ah well spotted. This was perhaps why when I tried to initially change the mode I was needing to use the idle command pad to prevent double entry

**Reply by netmindz** [reply]
Worth logging a bug with the AHA library if seeing a clash?

**Reply by netmindz** [reply]
Can't send multiple commands per pin5 btw as we only trigger after we read the FA message and that can be only once per pin5 cycle @tmjo

**Reply by tmjo** [reply]
You're probably right. Even if there were several loops per pin low, the buffer would be empty and not enter the if-statement I guess. It was just an idea I had, but probably not anything worth following up on then. 

I we conclude that the concept of 3-4 empty "paddings" works consistently, I guess that is jolly good, it just seems a bit weird. On the other hand, our panels send empty commands all the time and human interaction would certainly have a delay, so perhaps that's just the way it works.

**Reply by netmindz** [reply]
Does this do the same as you were trying? I've not tested yet, but hopefully matches what you are seeing @caldog20 
 @tmjo  https://github.com/netmindz/balboa_GL_ML_spa_control/pull/36

**Reply by tmjo** [reply]
I'll see if I can test tomorrow. Was going to send a PR by the way, but other stuff came in the way.

**Reply by netmindz** [reply]
I tested and it fails, but think that might be my own hardware setup
